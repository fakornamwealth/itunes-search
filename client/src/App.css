.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  color: white;
}

.App-link {
  color: #61dafb;
}

input,
select {
  display: block;
  width: 100%;
  padding: 0.6em;
  margin-bottom: 0.4em;
  box-sizing: border-box;
}

hr {
  width: 90%;
}

.item {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-bottom: 0.6em;
  width: 940px;
  border-bottom: solid 1px #fff;
  margin-bottom: 1em;
  text-align: left;
}

.item .item-info {
  flex-grow: 1;
}
.item .item-actions button {
  padding: 0.6em;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
