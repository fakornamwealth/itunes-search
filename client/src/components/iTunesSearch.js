// function to fetch search results from the backend
function iTunesSearch(term, media) {
  return fetch("http://localhost:3000/search", {
    method: "post", // post method is used to send search query parameters to the backend
    headers: {
      "Content-Type": "application/json;charset=utf-8",
    },
    body: JSON.stringify({
      // send search query parameters as a json formatted string
      term,
      media,
    }),
  })
    .then((response) => response.json())
    .then((response) => {
      //console.log(response);
      return response.results;
    });
}

export default iTunesSearch; // export function for usage and testing
