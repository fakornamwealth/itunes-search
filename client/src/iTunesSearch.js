function iTunesSearch(term, media) {
  return fetch("http://localhost:3000/search", {
    method: "post",
    headers: {
      "Content-Type": "application/json;charset=utf-8",
    },
    body: JSON.stringify({
      term,
      media,
    }),
  })
    .then((response) => response.json())
    .then((response) => {
      console.log(response);
      //setSearchResult(response.results);
      return response.results;
    });
}

export default iTunesSearch;
