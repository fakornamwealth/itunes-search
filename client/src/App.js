import { useState, useEffect } from "react";
import logo from "./logo.svg";
import "./App.css";

const mediaTypes = [
  {
    name: "Movie",
    value: "movie",
  },
  {
    name: "Podcast",
    value: "podcast",
  },
  {
    name: "Music",
    value: "music",
  },
  {
    name: "Music Video",
    value: "musicVideo",
  },
  {
    name: "Audiobook",
    value: "audiobook",
  },
  {
    name: "Short Film",
    value: "shortFilm",
  },
  {
    name: "TV Show",
    value: "tvShow",
  },
  {
    name: "Software",
    value: "software",
  },
  {
    name: "E-Book",
    value: "ebook",
  },
  {
    name: "All",
    value: "all",
  },
];

function App() {
  const [searchResult, setSearchResult] = useState([]);
  const [favs, setFavs] = useState([]);

  function iTunesSearch(event) {
    event.preventDefault();
    fetch("http://localhost:3000/search", {
      method: "post",
      headers: {
        "Content-Type": "application/json;charset=utf-8",
      },
      body: JSON.stringify({
        term: event.target.term.value,
        media: event.target.media.value,
      }),
    })
      .then((response) => response.json())
      .then((response) => {
        console.log(response);
        setSearchResult(response.results);
      });
  }

  function getFavs() {
    //const term = prompt("Enter search term:");
    fetch(`http://localhost:3000/favs`)
      .then((result) => result.json())
      .then((result) => {
        console.log(result);
        setFavs(result.favs);
      });
  }

  function addToFavourites(item) {
    console.log("Add to favourites...");
    //setFavs([...favs, item]);
    fetch(`http://localhost:3000/favs`, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ item }),
    })
      .then((result) => result.json())
      .then((result) => {
        console.log(result);
        setFavs(result.favs);
      });
  }

  function removeFav(item) {
    console.log("Remove from favourites...");
    //setFavs([...favs, item]);
    fetch(`http://localhost:3000/favs`, {
      method: "DELETE",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ item }),
    })
      .then((result) => result.json())
      .then((result) => {
        console.log(result);
        setFavs(result.favs);
      });
  }

  useEffect(() => {
    getFavs();
  }, []);

  return (
    <div className="App">
      <header className="App-header">
        <h1>iTunes Search</h1>
        <form onSubmit={iTunesSearch}>
          <input type="text" name="term" />
          <select name="media" defaultValue="all">
            {mediaTypes.map((mediaType) => {
              return (
                <option key={mediaType.value} value={mediaType.value}>
                  {mediaType.name}
                </option>
              );
            })}
          </select>
          <input type="submit" value="Search" />
        </form>
        <h2>Search Results</h2>
        <div id="results">
          {searchResult.map((result, index) => {
            return (
              <div className="item" key={index}>
                <div className="item-info">
                  {result.collectionName}, {result.artistName} ({result.kind})
                </div>
                <div className="item-actions">
                  <button
                    onClick={() => {
                      addToFavourites(result);
                      //console.log(favs);
                    }}
                  >
                    Add to favourites
                  </button>
                </div>
              </div>
            );
          })}
        </div>
        <hr />
        <h2>Favourites</h2>
        <div id="favs">
          {favs.map((item) => {
            return (
              <div key={item.collectionId} className="item">
                <div className="item-info">{item.artistName}</div>
                <div className="item-actions">
                  <button
                    onClick={() => {
                      console.log(item);
                      removeFav(item);
                    }}
                  >
                    Remove
                  </button>
                </div>
              </div>
            );
          })}
        </div>
      </header>
    </div>
  );
}

export default App;
