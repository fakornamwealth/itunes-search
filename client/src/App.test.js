// import dependencies
import renderer from "react-test-renderer";
import App from "./App";
import iTunesSearch from "./components/iTunesSearch";

// snapshot test
test("renders correctly", function () {
  let tree = renderer.create(<App />).toJSON(); // get json object representing the app
  expect(tree).toMatchSnapshot(); // compare to snapshot
});

// unit test
test("test fetch function", async function () {
  // overwrite native fetch function with a mock version
  // this will simulate a promise being returned with test data
  global.fetch = () =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          results: [
            {
              albumTitle: "Album Title", // test data
            },
          ],
        }),
    });

  // call our application's fetch function
  // the mock fetch function will be used instead of the native one
  const response = iTunesSearch("John", "all");

  //console.log(response);

  response.then((data) => {
    console.log(data);
    expect(data[0].albumTitle).toEqual("Album Title"); // compare result with test data
  });
});
