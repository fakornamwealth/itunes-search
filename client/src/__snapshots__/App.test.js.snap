// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <nav>
      <a
        href="/"
      >
        Home
      </a>
       | 
      <a
        href="/favs"
      >
        Favourites
      </a>
    </nav>
    <h1>
      iTunes Search
    </h1>
    <form
      onSubmit={[Function]}
    >
      <input
        name="term"
        type="text"
      />
      <select
        defaultValue="all"
        name="media"
      >
        <option
          value="movie"
        >
          Movie
        </option>
        <option
          value="podcast"
        >
          Podcast
        </option>
        <option
          value="music"
        >
          Music
        </option>
        <option
          value="musicVideo"
        >
          Music Video
        </option>
        <option
          value="audiobook"
        >
          Audiobook
        </option>
        <option
          value="shortFilm"
        >
          Short Film
        </option>
        <option
          value="tvShow"
        >
          TV Show
        </option>
        <option
          value="software"
        >
          Software
        </option>
        <option
          value="ebook"
        >
          E-Book
        </option>
        <option
          value="all"
        >
          All
        </option>
      </select>
      <input
        type="submit"
        value="Search"
      />
    </form>
    <h2>
      Search Results
    </h2>
    <div
      id="results"
    />
  </header>
</div>
`;
